design/rest-api-guideline.md

```diff
@@ -0,0 +1,388 @@
+* Be aware of extensive representations as they can affect overall performance and usability, if possible review your API to see if resources can be decomposed into smaller ones
+* Avoid deeply nested JSON objects
+* Avoid using generic resource representations
+
+### Http Status Codes
+
+The tables below defines the HTTP status code that clients should expect from APIs.
+
+This document also defines which codes are expected to be used for each HTTP verb, please take a look at [HTTP Verbs](#http-verbs) for more details.
+
+Clients should consider the status codes herein defined in order to properly handle responses from APIs, recover from errors, and 
+understand the reasoning behind the API responses.
+
+#### Success Codes
+
+|Verb|Description|
+|---|---|
+|200|Indicates the request completed successfully regardless the size of the response body|
+|201|Indicates the request completed successfully and a resource was created|
+|204|Indicates the request completed successfully and the server did not provide a response body|
+
+#### Redirection Codes
+
+|Verb|Description|
+|---|---|
+|307|Indicates that clients should resubmit the request to another location|
+
+#### Client Side Error Codes
+
+|Verb|Description|
+|---|---|
+|400|Indicates that the request is invalid, usually related with the validation of the payload|
+|401|Indicates that clients should provide authorization or the provided authorization is invalid|
+|403|Indicates that the authorization provided by the client is not enough to access the resource|
+|404|Indicates that the requested resource does not exist|
+|405|Indicates that the method chosen by the client to access a resource is not supported|
+|409|Indicates that the resource the client is trying to create already exists or some conflict when processing the request|
+
+#### Server Side Error Codes
+
+|Verb|Description|
+|---|---|
+|500|Indicates that the server could not fulfill the request due to some unexpected error|
+
+### Http Verbs
+
+Based on the REST architectural style, HTTP Verbs should be used as the means to communicate to APIs the operations to be performed
+on their resources. While HTTP verbs cover by themselves the most common operations you might want to look the [Controller Resource](#controllerresource) as an alternative on how to overcome any limitation imposed by the usage of HTTP verbs herein defined.
+
+The table below defines the HTTP verbs you should consider when exposing operations through your API and the circumstances where they apply.
+
+|Verb|Description|Idempotent|
+|---|---|---|
+|GET|Requests the current state of a resource|Yes|
+|POST|Creates a new resource or submit a command|No|
+|PUT|Replaces a resource state|Yes|
+|DELETE|Removes a resource|Yes|
+
+[TODO]: Shall we encourage PATCH for partial updates. That could make things easier for clients and avoid confusion when using PUT to update resources, where PUT
```

Comment:
+1 to include PATCH

Reply:
+1

Reply:
PATCH is a must! I would also like to see this guideline cover how we will support patch for JSON (perhaps looking at existing RFCs like https://tools.ietf.org/html/rfc7396, https://tools.ietf.org/html/rfc6901, https://tools.ietf.org/html/rfc6902).

Partial updates in the current admin endpoints is one of the messiest and inconsistent parts of the API. Sometimes it does partial updates, sometimes it doesn't. Sometimes it'll add to an array, sometimes it'll replace the array, etc..

Reply:
PATH is already in the TODO list.

I agree with you and that is the reason why I've left this open for discussion. PATH will help to communicate better and consistently how we perform updates together with `PUT`.

However, I don't think we should strictly follow those standards you mentioned, but something simpler where the payload sent using PATH does only partial updates. We are kind of doing RFC7396 already in some places, so we should stick with it. The other specs will make things more complicated, IMO. For both consumers and API developers.

Reply:
There is no clear contract in how we do partial updates today. We should not invent our own JSON patch approach, but rather stick to something already defined. There's even Java implementation of this stuff that allows you to do it fairly easily. What we have today is a confusing mess that is beyond inconsistent, so the alternative to using an existing spec is to invent our own, which is a bad idea.

Reply:
Sure, I'm not saying that there is a clear contract today. But that RFC7393 is what I think we should consider for partial updates. Which we do in some way in some of our APIs.